src/crypto/algorithms/megolm.js

```diff
@@ -283,104 +438,34 @@ MegolmEncryption.prototype._shareKeyWithDevices = function(session, devicesByUse
         this._olmDevice, this._baseApis, devicesByUser,
-    ).then(function(devicemap) {
-        const promises = [];
-
-        for (const userId in devicesByUser) {
-            if (!devicesByUser.hasOwnProperty(userId)) {
-                continue;
-            }
-
-            const devicesToShareWith = devicesByUser[userId];
-            const sessionResults = devicemap[userId];
-
-            for (let i = 0; i < devicesToShareWith.length; i++) {
-                const deviceInfo = devicesToShareWith[i];
-                const deviceId = deviceInfo.deviceId;
-
-                const sessionResult = sessionResults[deviceId];
-                if (!sessionResult.sessionId) {
-                    // no session with this device, probably because there
-                    // were no one-time keys.
-                    //
-                    // we could send them a to_device message anyway, as a
-                    // signal that they have missed out on the key sharing
-                    // message because of the lack of keys, but there's not
-                    // much point in that really; it will mostly serve to clog
-                    // up to_device inboxes.
-                    //
-                    // ensureOlmSessionsForUsers has already done the logging,
-                    // so just skip it.
-                    continue;
-                }
-
-                console.log(
-                    "sharing keys with device " + userId + ":" + deviceId,
-                );
-
-                const encryptedContent = {
-                    algorithm: olmlib.OLM_ALGORITHM,
-                    sender_key: self._olmDevice.deviceCurve25519Key,
-                    ciphertext: {},
-                };
+    );
 
-                if (!contentMap[userId]) {
-                    contentMap[userId] = {};
-                }
+    const userDeviceMaps = this._splitUserDeviceMap(
+        session, devicemap, devicesByUser,
+    );
 
-                contentMap[userId][deviceId] = encryptedContent;
-
-                promises.push(
-                    olmlib.encryptMessageForDevice(
-                        encryptedContent.ciphertext,
-                        self._userId,
-                        self._deviceId,
-                        self._olmDevice,
-                        userId,
-                        deviceInfo,
-                        payload,
-                    ),
-                );
-            }
-        }
+    if (userDeviceMaps.length == 0) {
+        // no devices to send to
+        return;
+    }
 
-        if (promises.length === 0) {
-            // no devices to send to
-            return Promise.resolve();
-        }
+    for (let i = 0; i < userDeviceMaps.length; i++) {
+        await this._encryptAndSendKeysToDevices(
+            session, userDeviceMaps[i], payload,
+        ).then(() => {
+            console.log(`Uploaded megolm keys in ${this._roomId} `
```

Comment:
"Uploaded" seems the wrong term. "Completed megolm keyshare" was more appropriate.